File: python/blog/sort/quickSort_hoare.py

def swap(arr, i, j):
    arr[i], arr[j] = arr[j], arr[i]

def partition(arr, le, ri):
    pivot = arr[(le + ri) // 2]
    i = le - 1
    j = ri + 1

    print('enter partition: ', arr, le, ri)

    while True:
        while True:
            i += 1
            print('i? ', i, pivot)
            if i >= ri or arr[i] >= pivot:
                break

        while True:
            j -= 1
            print('j? ', j, pivot)
            if j <= le or arr[j] <= pivot:
                break

        print('i: ', i, ', j: ', j)

        if i >= j:
            return j
        swap(arr, i, j)

def quickSort(arr, le, ri):
    if le >= ri:
        return

    j = partition(arr, le, ri)

    quickSort(arr, le, j)
    quickSort(arr, j + 1, ri)

#arr = [15, 20, 13, 30, 10, 25, 3, 24]
#arr = [5, 5, 5, 5, 5]
arr = [1, 2, 3, 4, 5]
print ('before quick sort: ', arr)
quickSort(arr, 0, len(arr) - 1)
print ('after quick sort: ', arr)

File: c/blog/sort/quickSort_hoare.c
#include <stdio.h>

void print(int* arr, int n)
{
	for (int i = 0; i < n; i++)
		printf("%d ", *arr++);
	printf("\n");
}

void swap(int* arr, int i, int j)
{
	int tmp = arr[i];
	arr[i] = arr[j];
	arr[j] = tmp;
}

int partition(int* arr, int left, int right)
{
	int pivot = arr[(left + right) / 2];
	int i = left - 1;
	int j = right + 1;

	while(1) {
		do {
			i++;
		} while(i < right && arr[i] < pivot);

		do {
			j--;
		} while(j > left && arr[j] > pivot);

		if (i >= j)
			return j;
		swap(arr, i, j);
	}
}

void quickSort(int* arr, int left, int right)
{
	if (left >= right)
		return;

	int j = partition(arr, left, right);

	quickSort(arr, left, j);
	quickSort(arr, j+1, right);
}

void main()
{
	int arr[] = {15, 20, 13, 30, 10, 25, 3, 24};
	int n = sizeof(arr) / sizeof(*arr);

	printf("before quick sort: ");
	print(arr, n);

	quickSort(arr, 0, n - 1);
	
	printf("after quick sort: ");
	print(arr, n);
}
